### DISCUSSION TITLE
Method argument types

### DISCUSSION THREAD

**Discussion by alejandro-zuleta-globant** [FreeOpcUa/opcua-asyncio#1293] [Q&A]
I have a method in my OPC UA server (`asyncua`) which gets a custom data type complex object as an argument. The method is _linked_ from address space which is populated using a custom nodeset XML file.

When I try to invoke the method from UAExpert I can't define the variables values (`Temperature` and `Units`) in the input argument.

![image](https://user-images.githubusercontent.com/108742827/234932923-aa9ca291-eec4-4adc-b5dd-92a37d02ee25.png)

There is only one plain-text field.

What is the recommended OPC UA structure for complex arguments?
How can I model that structure in UA nodeset?

This is a piece of my custom nodeset XML file with the related nodes:

```xml

<!-- MyArgument data type  -->

<UADataType NodeId="ns=2;i=7001" ParentNodeId="i=24" BrowseName="MyArgumentCommandType">
    <References>
      <Reference ReferenceType="HasSubtype" IsForward="false" TargetId="ns=2;i=7000" />
    </References>
    <UAVariable DataType="Double" BrowseName="Temperature" />
    <UAVariable DataType="String" BrowseName="Units" />
</UADataType>


<!-- MyMethod method  -->

<UAMethod BrowseName="MyMethod" NodeId="ns=2;i=6160" ParentNodeId="ns=2;i=6060">
    <DisplayName>MyMethod</DisplayName>
    <Description>Do x and y</Description>
    <References>
      <Reference IsForward="false" ReferenceType="HasComponent">ns=2;i=6060</Reference>
      <Reference ReferenceType="HasProperty">ns=2;i=6161</Reference>
      <Reference ReferenceType="HasProperty">ns=2;i=6162</Reference>
    </References>
</UAMethod>


<!-- MyMethod input arguments  -->


<UAVariable BrowseName="InputArguments" DataType="Argument" NodeId="ns=2;i=6161"
    ParentNodeId="ns=2;i=6160" ValueRank="0">
    <DisplayName>InputArguments</DisplayName>
    <Description>InputArguments</Description>
    <Value>
      <uax:ListOfExtensionObject>
        <uax:ExtensionObject>
          <uax:TypeId>
            <uax:Identifier>i=296</uax:Identifier>
          </uax:TypeId>
          <uax:Body>
            <uax:Argument>
              <uax:ValueRank>-1</uax:ValueRank>
              <uax:Description>
                <uax:Text />
                <uax:Locale />
              </uax:Description>
              <uax:DataType>
                <uax:Identifier>ns=2;i=7001</uax:Identifier>
              </uax:DataType>
              <uax:Name>data</uax:Name>
              <uax:ArrayDimensions />
            </uax:Argument>
          </uax:Body>
        </uax:ExtensionObject>
      </uax:ListOfExtensionObject>
    </Value>
    <References>
      <Reference IsForward="false" ReferenceType="HasProperty">ns=2;i=6090</Reference>
      <Reference ReferenceType="HasTypeDefinition">i=68</Reference>
    </References>
  </UAVariable>

<!-- MyMethod output arguments  -->


<UAVariable BrowseName="OutputArguments" DataType="Argument" NodeId="ns=2;i=6162"
    ParentNodeId="ns=2;i=6160" ValueRank="0">
    <DisplayName>OutputArguments</DisplayName>
    <Description>OutputArguments</Description>
    <Value>
      <uax:ListOfExtensionObject>
        <uax:ExtensionObject>
          <uax:TypeId>
            <uax:Identifier>i=296</uax:Identifier>
          </uax:TypeId>
          <uax:Body>
            <uax:Argument>
              <uax:ValueRank>-1</uax:ValueRank>
              <uax:Description>
                <uax:Text />
                <uax:Locale />
              </uax:Description>
              <uax:DataType>
                <uax:Identifier>i=12</uax:Identifier>
              </uax:DataType>
              <uax:Name>result</uax:Name>
              <uax:ArrayDimensions />
            </uax:Argument>
          </uax:Body>
        </uax:ExtensionObject>
      </uax:ListOfExtensionObject>
    </Value>
    <References>
      <Reference IsForward="false" ReferenceType="HasProperty">ns=2;i=6090</Reference>
      <Reference ReferenceType="HasTypeDefinition">i=68</Reference>
    </References>
</UAVariable>
```

Thanks in advance.

**Comment by danmoser** [comment]
You need to use asyncua's `Argument()` class.
> 
> https://github.com/FreeOpcUa/opcua-asyncio/blob/eabd784f05f2bbb8c177947c503d18d874982638/examples/server-extension-object-as-method-argument.py#L35
> 

See discussion here: https://github.com/FreeOpcUa/opcua-asyncio/discussions/852

**Reply by alejandro-zuleta-globant** [reply]
@danmoser , Thanks for your help. I need to model the method and arguments using a XML nodeset file instead of programmatically adding the nodes to the server via Python.

**Comment by schroeder-** [comment]
This is not valid UADatatype
```
<UADataType NodeId="ns=2;i=7001" ParentNodeId="i=24" BrowseName="MyArgumentCommandType">
    <References>
      <Reference ReferenceType="HasSubtype" IsForward="false" TargetId="ns=2;i=7000" />
    </References>
    <UAVariable DataType="Double" BrowseName="Temperature" />
    <UAVariable DataType="String" BrowseName="Units" />
</UADataType>
```
This is correct
```
<UADataType NodeId="ns=2;i=7001" ParentNodeId="i=24" BrowseName="MyArgumentCommandType">
    <References>
      <Reference ReferenceType="HasSubtype" IsForward="false" TargetId="ns=2;i=7000" />
    </References>
 <Field DataType="Double" Name="Temperature" />
    <Field DataType="String" Name="Units" />
</UADataType>
```

**Reply by plamenk** [reply]
Hi, I have a question re. this issue: I use link_method to implement a method in the server. It seems that load_data_type_definitions loads only the base UA Data Types under Structure, but not the additionally loaded Data Types from other nodesets. 
How can I use Data Types from imported xml nodesets, for example to create an output within my implemented method in the form:

...
    value = ua.uaprotocol_auto.F_JobResponseDataType()
    value.JobOrderID = input.Value
...

May data type, generated by SiOME under Types/DataTypes/BaseDataType/Structure is here:

    <UADataType NodeId="ns=10;i=3005" BrowseName="10:F_JobResponseDataType">
        <DisplayName>F_JobResponseDataType</DisplayName>
        <References>
            <Reference ReferenceType="HasSubtype" IsForward="false">i=22</Reference>
            <Reference ReferenceType="HasEncoding">ns=10;i=5092</Reference>
            <Reference ReferenceType="HasEncoding">ns=10;i=5093</Reference>
        </References>
        <Definition Name="F_JobResponseDataType">
            <Field DataType="String" Name="JobOrderID"/>
            <Field IsOptional="true" DataType="String" Name="Description"/>
            <Field IsOptional="true" DataType="String" Name="ProductID"/>
            <Field IsOptional="true" DataType="String" Name="PartSerialNumber"/>
            <Field IsOptional="true" DataType="DateTime" Name="StartTime"/>
            <Field IsOptional="true" DataType="DateTime" Name="EndTime"/>
            <Field DataType="ISA95JobOrderStateEnum" Name="JobState"/>
            <Field IsOptional="true" DataType="ns=10;i=3013" Name="PartState"/>
            <Field IsOptional="true" DataType="Int32" Name="LinePosition"/>
            <Field IsOptional="true" DataType="String" Name="MaterialContainerIDIn"/>
            <Field IsOptional="true" DataType="String" Name="MaterialContainerIDOut"/>
            <Field IsOptional="true" DataType="ns=10;i=3012" Name="JobResponseData"/>
            <Field IsOptional="true" DataType="ns=10;i=3017" ValueRank="1" Name="PersonnelActuals"/>
            <Field IsOptional="true" DataType="ns=10;i=3015" ValueRank="1" Name="EquipmentActuals"/>
            <Field IsOptional="true" DataType="ns=10;i=3018" ValueRank="1" Name="PhysicalAssetActuals"/>
            <Field IsOptional="true" DataType="ns=10;i=3001" ValueRank="1" Name="MaterialActuals"/>
        </Definition>
    </UADataType>

Thanks.

Plamen
